qa-tests-backend/src/main/groovy/util/Retry.groovy

```diff
@@ -0,0 +1,97 @@
+
+import groovy.transform.CompileStatic
+import org.codehaus.groovy.ast.ASTNode
+import org.codehaus.groovy.ast.AnnotationNode
+import org.codehaus.groovy.ast.ClassHelper
+import org.codehaus.groovy.ast.MethodNode
+import org.codehaus.groovy.ast.Parameter
+import org.codehaus.groovy.ast.VariableScope
+import org.codehaus.groovy.ast.expr.ArgumentListExpression
+import org.codehaus.groovy.ast.expr.ClosureExpression
+import org.codehaus.groovy.ast.expr.ConstantExpression
+import org.codehaus.groovy.ast.expr.MethodCallExpression
+import org.codehaus.groovy.ast.expr.StaticMethodCallExpression
+import org.codehaus.groovy.ast.expr.VariableExpression
+import org.codehaus.groovy.ast.stmt.ReturnStatement
+import org.codehaus.groovy.control.SourceUnit
+import org.codehaus.groovy.transform.AbstractASTTransformation
+import org.codehaus.groovy.transform.GroovyASTTransformation
+import org.codehaus.groovy.transform.GroovyASTTransformationClass
+
+/**
+ * Retry the method in case of exception. Using Helpers.evaluateWithRetry
+ */
+@Documented
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.METHOD)
+@GroovyASTTransformationClass(classes = RetryASTTransformation)
+@interface Retry {
+
+    /**
+     * How many times to retry.
+     * @return Number of attempts
+     */
+    int attempts() default 3
+
+    /**
+     * Delay in seconds between attempts, in time units.
+     * @return Delay
+     */
+    long delay() default 1
+}
+
+@GroovyASTTransformation
+class RetryASTTransformation extends AbstractASTTransformation {
+
+    @Override
+    @CompileStatic
+    void visit(ASTNode[] nodes, SourceUnit sourceUnit) {
+        AnnotationNode annotation = (AnnotationNode) nodes[0]
+        MethodNode method = (MethodNode) nodes[1]
+
+        def clazz = method.declaringClass
+        def methodName = method.name + "_with_retry"
+        clazz.addMethod(
+                methodName,
+                method.modifiers,
+                method.returnType,
+                method.parameters,
+                method.exceptions,
+                method.code
+        )
+
+        int attempts = getMemberIntValue(annotation, "attempts")
+        int delay = getMemberIntValue(annotation, "delay")
+
+        def argumentListExpression = new ArgumentListExpression(method.parameters)
+        def funcCall = new MethodCallExpression(new VariableExpression("this"), methodName, argumentListExpression)
+
+        def closureExpression = new ClosureExpression(Parameter.EMPTY_ARRAY, new ReturnStatement(funcCall))
+        def variableScope = new VariableScope()
+        method.parameters.each {
+            variableScope.putReferencedLocalVariable(it)
+            it.setClosureSharedVariable(true)
+        }
+        closureExpression.setVariableScope(variableScope)
+
+        def retryCall = new ReturnStatement(
+                new StaticMethodCallExpression(
+                        ClassHelper.make(Helpers),
+                        "evaluateWithRetry",
```

Comment:
May we please implement some logging for when errors occur?
Similar messages to which are removed in the follow-up #14657.

Reply:
We already have logging in in `evaluateWithRetry`:
https://github.com/stackrox/stackrox/blob/d1247121590a1f25db09ce4e97b81bf884527495/qa-tests-backend/src/main/groovy/util/Helpers.groovy#L29

Reply:
Maybe it's not the best example but this is how it works in `RetryTest`
```
17:59:02 | DEBUG |                           | Helpers                   | Caught exception. Retrying in 0s (attempt 0 of 3): Condition not satisfied:

i++ > 2
|   |
0   false

I was called 0

17:59:02 | DEBUG |                           | Helpers                   | Caught exception. Retrying in 0s (attempt 1 of 3): Condition not satisfied:

i++ > 2
|   |
1   false

I was called 1

17:59:02 | DEBUG |                           | Helpers                   | Caught exception. Retrying in 0s (attempt 2 of 3): Condition not satisfied:

i++ > 2
|   |
2   false

I was called 2

> Task :test
RetryTest > test PASSED
```

Reply:
Right, but it might be not clear what is being retried, unless the test adds some logging. In your example, I'd prefer to see the `fail` function name somewhere. I guess you get this information from the introspection.

Please see the removed logging from the follow-up PR for a reference.